### PR TITLE
SI-8512 Infer a type for f"$args"

### PR DESCRIPTION
**Pull request by som-snytt**
The f-interpolator gets a type param that better be Any to avoid
unfortunate widenings.

Hey, it worked!

@densh 's comments on the ticket are longer than the fix and the test.

Review by @densh

### CODE REVIEW

File: src/library/scala/StringContext.scala [hunk 1]
```diff
@@ -163,7 +163,7 @@ case class StringContext(parts: String*) {
    */
   // The implementation is hardwired to `scala.tools.reflect.MacroImplementations.macro_StringInterpolation_f`
   // Using the mechanism implemented in `scala.tools.reflect.FastTrack`
-  def f(args: Any*): String = macro ???
+  def f[A >: Any](args: A*): String = macro ???
```
Comment (xeno-by) [inline src/library/scala/StringContext.scala:166]: Which kind of widenings does this prevent? Could you elaborate a bit more?

Comment (som-snytt) [inline src/library/scala/StringContext.scala:166]: In the unit test, f"${3.14}%.2f rounds to ${3}%d", `undetParam inferred: type A as Double` and 

```
performing macro expansion scala.StringContext.apply("", "%f rounds to ", "%d").f[Double](3.14, 3.0)
```

Which I take to mean the 3 becomes 3.0 before the macro sees it.  I haven't checked yet for anything that enables undoing that conversion (like an attachment or history of some kind).  Maybe a lazy Sunday afternoon is a good time to read some code.

**Comment by som-snytt** [comment]
Jenkins couldn't find 2.10.2.  Who uses 2.10.2 anymore?  Moreover, Jenkins was unable to send an alert email to typesafe to let them know.  Doesn't Jenkins know that typesafe takes August off?  Since Roman times, August has been known as the least typesafe month of the year.

File: src/library/scala/StringContext.scala (hunk 1 above)
Comment (densh) [inline src/library/scala/StringContext.scala:166]: That's quite surprising. It turns out that quasiquotes have the same issue here:

```
scala> q"${1} + ${1.0}"
res1: reflect.runtime.universe.Tree = 1.0.$plus(1.0)
```

Please add this bound to quasiquote interpolator definition too https://github.com/scala/scala/blob/2.12.x/src/reflect/scala/reflect/api/Quasiquotes.scala#L16

**Comment by gourlaysama** [comment]
PLS REBUILD/pr-scala@450e78a

**Comment by gourlaysama** [comment]
Maven central seems to be fixed, let's see if that works...

File: src/library/scala/StringContext.scala (hunk 1 above)
Comment (som-snytt) [inline src/library/scala/StringContext.scala:166]: No problemo.

**Comment by gkossakowski** [comment]
PLS REBUILD ALL

**Comment by som-snytt** [comment]
Rebased, conflict at added test functions. Now that's something it could figure out with minimal smarts.

**Comment by densh** [comment]
LGTM

**Comment by gkossakowski** [comment]
Thanks!

**Pull request merged**